src/alpaca_typer.erl

```diff
@@ -1606,28 +1640,37 @@ typ_of(Env, Lvl, #alpaca_apply{line=L, expr=Expr, args=Args}) ->
                         case typ_of(Env, Lvl, Fun) of
                             {error, _}=Err -> Err;
                             {TypF, NextVar} ->
-                                typ_apply(Env, Lvl, TypF, NextVar, Args, L)
+                                %% We should have a t_arrow taking some args with a return value
+                                %% What we need is a t_arrow that takes some of those args and returns
+                                %% another t_arrow taking the remainder and returning the final arg
+                                try
+                                    typ_apply(Env, Lvl, TypF, NextVar, Args, L)
+                                catch
+                                    error:{arity_error, _, _} = Err -> CurryFun(Err)
```

Comment:
Since `{arity_error, _, _}` is module internal and used for control flow, it is more idiomatic to use `throw` rather than `error`.